src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -235,180 +273,215 @@ protected override AuthenticationTicket AuthenticateCore()
+                        new SecurityTokenValidatedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+                        {
+                            AuthenticationTicket = ticket,
+                            ProtocolMessage = message
+                        };
+
+                    await Options.Notifications.SecurityTokenValidated(securityTokenValidatedNotification);
+                    if (securityTokenValidatedNotification.HandledResponse)
                     {
-                        ticket.Properties.ExpiresUtc = expires;
+                        _logger.LogInformation(message: Resources.OIDCH_0012_SecurityTokenValidatedNotificationHandledResponse);
+                        return securityTokenValidatedNotification.AuthenticationTicket;
                     }
 
-                    ticket.Properties.AllowRefresh = false;
-                }
+                    if (securityTokenValidatedNotification.Skipped)
+                    {
+                        _logger.LogInformation(message: Resources.OIDCH_0013_SecurityTokenValidatedNotificationSkipped);
+                        return null;
+                    }
 
-                var securityTokenValidatedNotification = new SecurityTokenValidatedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
-                {
-                    AuthenticationTicket = ticket,
-                    ProtocolMessage = openIdConnectMessage
-                };
+                    string nonce = jwt.Payload.Nonce;
+                    if (Options.NonceCache != null)
+                    {
+                        // if the nonce cannot be removed, it was used
+                        if (!Options.NonceCache.TryRemoveNonce(nonce))
+                        {
+                            nonce = null;
+                        }
+                    }
+                    else
+                    {
+                        nonce = ReadNonceCookie(nonce);
+                    }
 
-                await Options.Notifications.SecurityTokenValidated(securityTokenValidatedNotification);
-                if (securityTokenValidatedNotification.HandledResponse)
-                {
-                    return securityTokenValidatedNotification.AuthenticationTicket;
-                }
+                    var protocolValidationContext = new OpenIdConnectProtocolValidationContext
+                    {
+                        AuthorizationCode = message.Code,
+                        Nonce = nonce, 
+                    };
 
-                if (securityTokenValidatedNotification.Skipped)
-                {
-                    return null;
+                    Options.ProtocolValidator.Validate(jwt, protocolValidationContext);
                 }
 
-                var protocolValidationContext = new OpenIdConnectProtocolValidationContext
+                if (message.Code != null)
```

Comment:
Have you considered adding a built-in HTTP client to automagically retrieve an access token using the authorization code? It would be really useful.